### PR TITLE
When creating elastic search index, the index that can delete histori…

### PR DESCRIPTION
**Pull request by yeerpeng**
When creating elastic search index, the index that can delete historical data can be created by day by configuration control. When deleting historical data, the index can be deleted directly, which improves the performance of elastic search.

Please answer these questions before submitting pull request

- Why submit this pull request?
New feature provided

- Related issues

___
### New feature or improvement
- Describe the details and related test reports.
When creating elastic search index, the index that can delete historical data can be created by day by configuration control. When deleting historical data, the index can be deleted directly, which improves the performance of elastic search.
The TraceQueryService.queryTrace method has changed in the UI layer. See the UI submission for details.

### CODE REVIEW

**Comment by wu-sheng** [comment]
I noticed https://github.com/apache/incubator-skywalking-ui/pull/236 . Why do you need to change query in this implicit? This is bad for integration. 

We have another UI, RocketBot, we are discussing to accept it as primary UI. This change could break it.
cc @TinyAllen

**Comment by wu-sheng** [comment]
If query protocol needs to be changed, it must be changed in protocol level.

**Comment by yeerpeng** [comment]
Daily index queries need time to determine which index to route to, but the TraceQueryService. queryTrace (String traceId) method has no timestamp

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java [hunk 1]
```diff
@@ -63,6 +64,28 @@ public void start(ModuleManager moduleManager) {
...
+        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(new RunnableWithExceptionProtection(this::deleteAndCreateOther, t -> logger.error("delete index in background failure.", t)), 1, 2, TimeUnit.HOURS);
+    }
+
+    private void deleteAndCreateOther() {
+        TimeBuckets timeBuckets = convertTimeBucket(new DateTime());
+        IModelGetter modelGetter = moduleManager.find(CoreModule.NAME).provider().getService(IModelGetter.class);
+        List<Model> models = modelGetter.getModels();
+        DownsamplingConfigService downsamplingConfigService = moduleManager.find(CoreModule.NAME).provider().getService(DownsamplingConfigService.class);
+        List<Model> allModels = new ArrayList<>();
+        models.forEach(model -> {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: Create so many indexes daily? Look like crazy. I think only segment index has performance concern.

**Comment by wu-sheng** [comment]
cc @peng-yongsheng Welcome to join this review. I think you mentioned someone else has a segment index change too.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java (hunk 1 above)
Comment (yeerpeng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: The index that is created on a daily basis is determined in the application.yml configuration file. Not all indexes are created that way

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: That makes sense. But, I still think, you could provide this feature as a part of es storage implementation only. Not in the core.

Comment (yeerpeng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: Well, yes, you may not have seen carefully the changes I made. The modifications in core are only historical data deletion operations. Specific changes in data storage and data search are all modified in the module storage-elasticsearch-plugin

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: Delete is one of the concerns too. If that delete API changed, it means all storage implementors(such as H2, MySQL) need to understand the ES index things, which make the API change. But actually, they may don't need to know.

You could think in this way, if I say the index created by daily is another form of storage implementor, how do you do the changes?

The best way is, you could deal with the delete inside the existing delete method. 

This made my another concern, `query/TraceQueryService.java`. You changed query too, which is much harder to be understood by frontend developer and others do the integration.

If you could support all indexes running by daily,  it is cool, but consider this as an advanced feature. Or, we could say, a new `elasticsearch-large-scale` provider, based on the existing one.

Comment (yeerpeng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: Oh, I totally understand what you mean. First of all, thanks for your patient explanation. I think I know what to do.

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: Look forward your update :)

`elasticsearch-large-scale` looks like really reasonable to me. In there, you could change settings, and make them more reasonable and clear, also reuse the existing DAO implementation.

Comment (yeerpeng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/ttl/DataTTLKeeperTimer.java]: Yes, I can make great progress in this way, and I am also willing to do such a thing :)

**Comment by peng-yongsheng** [comment]
@yeerpeng @wu-sheng 
1. From the previous pressure measurement, the record insertion and deletion are the key issues. There is no need to change the metrics. Why IndicatorEsDAO has been changed. 
2. There is no need to add a new method named historyOther in IHistoryDeleteDAO, the parameters in deleteHistory method are enough to delete the index by the time.
3. No found the implementation of the other DBs.
4. Just for record, Why are other queries affected?  getClient().search(new String[]{EndpointInventory.MODEL_NAME}, searchSourceBuilder)
5. queryTrace by trace id is really needs the UI to be modified or we can try to search from the records indexes. Not too more, and the trace id is primary key. So I think the performance loss is acceptable.

**Comment by wu-sheng** [comment]
> queryTrace by trace id is really needs the UI to be modified or we can try to search from the records indexes. Not too more, and the trace id is primary key. So I think the performance loss is acceptable.

This is hard to say. Which do you expect UI knows? Or which field should we add into GraphQL protocol?

**Pull request closed without merging**